dot/parachain/bitfield-distribution/bitfield_distribution.go

```diff
@@ -0,0 +1,533 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package bitfielddistribution
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/dot/parachain/grid"
+	networkbridgeevents "github.com/ChainSafe/gossamer/dot/parachain/network-bridge/events"
+	networkbridgemessages "github.com/ChainSafe/gossamer/dot/parachain/network-bridge/messages"
+	provisionermessages "github.com/ChainSafe/gossamer/dot/parachain/provisioner/messages"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/dot/parachain/util"
+	validationprotocol "github.com/ChainSafe/gossamer/dot/parachain/validation-protocol"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-bitfield-distribution"))
+
+type perRelayParentData struct {
+	// Signing context for a particular relay parent.
+	// the required part of the signing context
+	sessionIndex parachaintypes.SessionIndex
+
+	// Set of validators for a particular relay parent.
+	validatorsSet []parachaintypes.ValidatorID
+
+	// Set of validators for a particular relay parent for which we
+	// received a valid `BitfieldGossipMessage`.
+	// Also serves as the list of known messages for peers connecting
+	// after bitfield gossips were already received.
+	onePerValidator map[parachaintypes.ValidatorID]*validationprotocol.BitfieldDistributionMessage
+
+	// Avoid duplicate message transmission to our peers.
+	messageSentToPeer map[peer.ID]map[parachaintypes.ValidatorID]struct{}
+
+	// Track messages that were already received by a peer to prevent flooding.
+	messageReceivedFromPeer map[peer.ID]map[parachaintypes.ValidatorID]struct{}
+}
+
+// this will be used in the PeerViewChange handler so skip the lint check for now
+//
+//nolint:all
+func newPerRelayParentData(sessionIndex parachaintypes.SessionIndex, validatorSet []parachaintypes.ValidatorID,
+) *perRelayParentData {
+	return &perRelayParentData{
+		sessionIndex:            sessionIndex,
+		validatorsSet:           validatorSet,
+		onePerValidator:         make(map[parachaintypes.ValidatorID]*validationprotocol.BitfieldDistributionMessage),
+		messageSentToPeer:       make(map[peer.ID]map[parachaintypes.ValidatorID]struct{}),
+		messageReceivedFromPeer: make(map[peer.ID]map[parachaintypes.ValidatorID]struct{}),
+	}
+}
+
+// messageFromValidatorNeededByPeer determines if that particular message signed by a
+// validator is needed by the given peer.
+func (p *perRelayParentData) messageFromValidatorNeededByPeer(peerID peer.ID, signedBy parachaintypes.
+	ValidatorID) bool {
+	_, sendToExist := p.messageSentToPeer[peerID][signedBy]
+	_, receiveFromExist := p.messageReceivedFromPeer[peerID][signedBy]
```

Comment:
if the peerID does not exists in the first map it will return nil and search for `signedBy` in a nil map can cause a panic, right? should we check the first map retrival?

Reply:
that is a valid concern man, will change that